openedx/core/djangoapps/content/course_structures/tasks.py

```diff
@@ -0,0 +1,79 @@
+import json
+import logging
+
+from celery.task import task
+from opaque_keys.edx.keys import CourseKey
+from xmodule.modulestore.django import modulestore
+
+
+log = logging.getLogger(__name__)
+
+
+def generate_course_structure(course_key):
+    """
+    Generates a course structure dictionary for the specified course.
+    """
+    course = modulestore().get_course(course_key, depth=None)
+    blocks_stack = [course]
+    blocks_dict = {}
+    while blocks_stack:
+        curr_block = blocks_stack.pop()
+        children = curr_block.get_children() if curr_block.has_children else []
+        key = unicode(curr_block.scope_ids.usage_id)
+        block = {
+            "usage_key": key,
+            "block_type": curr_block.category,
+            "display_name": curr_block.display_name,
+            "children": [unicode(child.scope_ids.usage_id) for child in children]
+        }
+
+        # Retrieve these attributes separately so that we can fail gracefully if the block doesn't have the attribute.
+        attrs = (('graded', False), ('format', None))
+        for attr, default in attrs:
+            try:
+                block[attr] = getattr(curr_block, attr, default)
+            except Exception:
+                log.exception('Failed to retrieve %s attribute of block %s. Defaulting to %s.', attr, key, default)
```

Comment:
The `__get__` method used to retrieve the attribute is the source of the exception here. Your proposed change will continue to raise an exception. I've included a stack trace if you want to dive deeper:

```
Traceback (most recent call last):
    openedx/core/djangoapps/content/course_structures/tests.py line 92 in test_block_with_missing_fields
      structure = _generate_course_structure(self.course.id)
    openedx/core/djangoapps/content/course_structures/tasks.py line 33 in _generate_course_structure
      block[attr] = getattr(curr_block, attr, default)
    common/lib/xmodule/xmodule/x_module.py line 575 in __get__
      return getattr(getattr(instance, self._source), self._name)
    common/lib/xmodule/xmodule/x_module.py line 1034 in _xmodule
      raise UndefinedContext()
   UndefinedContext: 
```